### PR TITLE
Missing data

### PR DESCRIPTION
**Pull request by tcoenen**

### CODE REVIEW

File: frontend/src/components/DataSummary.vue [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+
+<script>
+import { readPaginatedData } from '@/services/datareader'
+import _ from 'lodash'
+
+const URL = process.env.API_HOST + '/onderwijs/api/data-summary/'
+const nYears = 4
+
+function getFullHeader (rows) {
+  let fullHeader = rows.reduce((header, currentRow, idx) => {
```
Comment (Vegter) [inline frontend/src/components/DataSummary.vue]: I would suggest removing the {} and write the reduce as a "one-liner"

Comment (Vegter) [reply frontend/src/components/DataSummary.vue]: Remove obsolete comments

File: frontend/src/components/DataSummary.vue [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+      }
+    },
+    setVariableOrder () {
+      if (this.header) {
+        // sort the variables (aka. available datasets) alphabetically
+        // then sort the entries per variabel numerically (in case of years) and convert to array
+        // make sure the array does not contain the true values
+        let ordered = Object.entries(this.header)
+          .sort((a, b) => a[0] > b[0])
+          .map(d => ({
```
Comment (Vegter) [inline frontend/src/components/DataSummary.vue]: Can't you use destructuring to circumvent magic values like d[0] etc? something like
[variables, entries] => ...

File: frontend/src/components/DataSummary.vue [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+            variable: d[0],
+            entries: Object.entries(d[1])
+              .sort((a, b) => a[0] > b[0])
+              .map(d => d[0])
+          }))
+        this.ordered = ordered
+      }
```
Comment (Vegter) [inline frontend/src/components/DataSummary.vue]: why not assign to this.ordered immediately?

File: frontend/src/components/DataSummary.vue [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+    }
+  },
+  watch: {
+    rows (to) {
+      this.setHeader(to)
+    },
+    header (to) {
+      if (to) {
+        console.log('HEADER', to)
+        this.setVariableOrder(to)
```
Comment (Vegter) [inline frontend/src/components/DataSummary.vue]: Remove debugging log

File: frontend/src/components/vestigingscijfers/CitoScore.vue [hunk 5]
```diff
@@ -1,27 +1,34 @@
...
+          )
+          : '...'
         }}
+      </h1>
+      <div class="average">
+        <div style="text-align: right">Amsterdams gemiddelde: {{
+          this.citoScoreGem !== null
+          ? this.citoScoreGem.toLocaleString(
+              'nl-DU', { maximumFractionDigits: 1 }
+          )
```
Comment (Vegter) [inline frontend/src/components/vestigingscijfers/CitoScore.vue]: You could make this a component (code reuse)

File: frontend/src/components/vestigingscijfers/CitoScore.vue [hunk 6]
```diff
@@ -34,18 +41,21 @@ export default {
...
     this.getData()
   },
   methods: {
     async getData () {
-      let score = await readPaginatedData(API_HOST + `/onderwijs/api/cito-score/?vestiging=${this.id}&jaar=2016`)
-      this.citoScoreData = score[0].cet_gem
-      let scoreGem = await getBbgaVariables(['OCITOSCH_GEM'], ['STAD'], [2016])
-      this.citoScoreGem = scoreGem[0].waarde
+      this.citoScoreData = _.get(await readPaginatedData(API_HOST + `/onderwijs/api/cito-score/?vestiging=${this.id}&jaar=2016`), '[0].cet_gem', null)
+      this.citoScoreGem = _.get(await getBbgaVariables(['OCITOSCH_GEM'], ['STAD'], [2016]), '[0].waarde', null)
```
Comment (Vegter) [inline frontend/src/components/vestigingscijfers/CitoScore.vue:54]: Is this really 2016?

Comment (Vegter) [reply frontend/src/components/vestigingscijfers/CitoScore.vue:54]: What happens in the future?

File: frontend/src/components/vestigingscijfers/CitoScore.vue [hunk 7]
```diff
@@ -34,18 +41,21 @@ export default {
...
+
+      if (!this.citoScoreData) {
+        this.MISSING_DATA = true
+      }
     }
```
Comment (Vegter) [inline frontend/src/components/vestigingscijfers/CitoScore.vue]: Is MISSING_DATA needed? Looks like just an alias for !cityScoreData

File: frontend/src/components/vestigingscijfers/leerlingennaargewicht/VestigingscijferLeerlingenNaarGewicht.vue [hunk 8]
```diff
@@ -36,6 +37,10 @@ export default {
       if (this.id) {
         let url = API_HOST + `/onderwijs/api/leerling-naar-gewicht/?vestiging=${this.id}`
         this.data = await readPaginatedData(url)
+
+        if (!this.data.length) {
+          this.MISSING_DATA = true
+        }
```
Comment (Vegter) [inline frontend/src/components/vestigingscijfers/leerlingennaargewicht/VestigingscijferLeerlingenNaarGewicht.vue]: Same here for MISSING_DATA

File: frontend/src/components/vestigingscijfers/schooladviezen/VestigingscijferSchoolAdvies.vue [hunk 9]
```diff
@@ -36,11 +39,13 @@ export default {
...
         let url = API_HOST + `/onderwijs/api/aggregated-advies/?vestiging=${this.id}`
-
-        let data = await readData(url)
-        order(data, 'advies', ['pro', 'vmbo b / k', 'vmbo g / t', 'havo / vwo'])
+        let data = order(await readData(url), 'advies', ['pro', 'vmbo b / k', 'vmbo g / t', 'havo / vwo'])
 
         this.data = data
+
+        if (!data.length) {
+          this.MISSING_DATA = true
```
Comment (Vegter) [inline frontend/src/components/vestigingscijfers/schooladviezen/VestigingscijferSchoolAdvies.vue]: Same here for MISSING_DATA

File: frontend/src/components/vestigingscijfers/schooladviezen/VestigingscijferSchoolAdvies.vue [hunk 10]
```diff
@@ -36,11 +39,13 @@ export default {
     async getData () {
       if (this.id) {
         let url = API_HOST + `/onderwijs/api/aggregated-advies/?vestiging=${this.id}`
-
-        let data = await readData(url)
-        order(data, 'advies', ['pro', 'vmbo b / k', 'vmbo g / t', 'havo / vwo'])
+        let data = order(await readData(url), 'advies', ['pro', 'vmbo b / k', 'vmbo g / t', 'havo / vwo'])
```
Comment (Vegter) [inline frontend/src/components/vestigingscijfers/schooladviezen/VestigingscijferSchoolAdvies.vue]: SCHOOL_TYPES in config somewhere?

File: web/app/dataset/create_summaries.py [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+        objects = list(queryset)
+        for obj in objects:
+            _key = '{}{:02d}'.format(obj.brin, obj.vestigingsnummer)
+            summary = summaries[_key][model.__name__] if model.__name__ in summaries[_key] else {}
+
+            if is_yearly:
+                entry = {obj.jaar: True}
+            else:
+                entry = {'present': True}
+            summary.update(entry)
```
Comment (Vegter) [inline web/app/dataset/create_summaries.py]: Why not use the same = if else as above?

File: web/app/dataset/create_summaries.py [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+            model_instances = list(
+                model.objects.filter(brin__exact=data_summary.row_key[:4]))
+            if not model_instances:
+                continue
+
+            s = data_summary.columns[model.__name__] \
+                if model.__name__ in data_summary.columns \
+                else {}
+
+            for obj in model_instances:
```
Comment (Vegter) [inline web/app/dataset/create_summaries.py:71]: Same here

**Review by Vegter — CHANGES_REQUESTED** [review]

File: frontend/src/components/DataSummary.vue (hunk 3 above)
Comment (tcoenen) [reply frontend/src/components/DataSummary.vue]: ok

File: frontend/src/components/DataSummary.vue (hunk 1 above)
Comment (tcoenen) [reply frontend/src/components/DataSummary.vue]: ok

File: frontend/src/components/DataSummary.vue (hunk 4 above)
Comment (tcoenen) [reply frontend/src/components/DataSummary.vue]: ok

File: frontend/src/components/vestigingscijfers/CitoScore.vue (hunk 6 above)
Comment (tcoenen) [reply frontend/src/components/vestigingscijfers/CitoScore.vue:54]: I will tackle that problem, but not in this ticket which is about missing data and reporting about it. (OP-199)

File: web/app/dataset/create_summaries.py (hunk 11 above)
Comment (tcoenen) [reply web/app/dataset/create_summaries.py]: I will do it for consistency, ternary expressions in Python are not as common as in Javascript though.

File: web/app/dataset/create_summaries.py (hunk 12 above)
Comment (tcoenen) [reply web/app/dataset/create_summaries.py:71]: idem

File: frontend/src/components/vestigingscijfers/CitoScore.vue (hunk 7 above)
Comment (tcoenen) [reply frontend/src/components/vestigingscijfers/CitoScore.vue]: There are three relevant states:
1) Data is present and downloaded -> show it.
2) APIs work, data is not available for this school -> notify user that we do not have the data.
3) There is some other problem in the APIs (slow, offline) -> show the elipsis.

One boolean cannot encode 3 states hence the current solution.

File: frontend/src/components/vestigingscijfers/leerlingennaargewicht/VestigingscijferLeerlingenNaarGewicht.vue (hunk 8 above)
Comment (tcoenen) [reply frontend/src/components/vestigingscijfers/leerlingennaargewicht/VestigingscijferLeerlingenNaarGewicht.vue]: idem

File: frontend/src/components/vestigingscijfers/schooladviezen/VestigingscijferSchoolAdvies.vue (hunk 9 above)
Comment (tcoenen) [reply frontend/src/components/vestigingscijfers/schooladviezen/VestigingscijferSchoolAdvies.vue]: idem

File: frontend/src/components/DataSummary.vue (hunk 2 above)
Comment (tcoenen) [reply frontend/src/components/DataSummary.vue]: ok

File: frontend/src/components/vestigingscijfers/schooladviezen/VestigingscijferSchoolAdvies.vue (hunk 10 above)
Comment (tcoenen) [reply frontend/src/components/vestigingscijfers/schooladviezen/VestigingscijferSchoolAdvies.vue]: will look into it, not relevant to OP-199

File: frontend/src/components/vestigingscijfers/CitoScore.vue (hunk 7 above)
Comment (Vegter) [reply frontend/src/components/vestigingscijfers/CitoScore.vue]: Well, if it is a state then make it a state I would say

Comment (tcoenen) [reply frontend/src/components/vestigingscijfers/CitoScore.vue]: we talked about this, have implemented as agreed

File: frontend/src/components/vestigingscijfers/leerlingennaargewicht/VestigingscijferLeerlingenNaarGewicht.vue (hunk 8 above)
Comment (tcoenen) [reply frontend/src/components/vestigingscijfers/leerlingennaargewicht/VestigingscijferLeerlingenNaarGewicht.vue]: changed to the hasData style

**Pull request merged**